crates/ide-diagnostics/src/handlers/json_is_not_rust.rs

```diff
@@ -0,0 +1,201 @@
+//! This diagnostic provides an assist for creating a struct definition from a json
+//! example.
+
+use ide_db::{base_db::FileId, source_change::SourceChange};
+use itertools::Itertools;
+use stdx::format_to;
+use syntax::{
+    ast::{self, make},
+    SyntaxKind, SyntaxNode,
+};
+use text_edit::TextEdit;
+
+use crate::{fix, Diagnostic, Severity};
+
+#[derive(Default)]
+struct State {
+    result: String,
+    struct_counts: usize,
+}
+
+impl State {
+    fn generate_new_name(&mut self) -> ast::Name {
+        self.struct_counts += 1;
+        make::name(&format!("Struct{}", self.struct_counts))
+    }
+
+    fn build_struct(&mut self, value: &serde_json::Map<String, serde_json::Value>) -> ast::Type {
+        let name = self.generate_new_name();
+        let ty = make::ty(&name.to_string());
+        let strukt = make::struct_(
+            None,
+            name,
+            None,
+            make::record_field_list(value.iter().sorted_unstable_by_key(|x| x.0).map(
+                |(name, value)| make::record_field(None, make::name(name), self.type_of(value)),
+            ))
+            .into(),
+        );
+        format_to!(self.result, "#[derive(Serialize, Deserialize)]\n{}\n", strukt);
+        ty
+    }
+
+    fn type_of(&mut self, value: &serde_json::Value) -> ast::Type {
+        match value {
+            serde_json::Value::Null => make::ty_unit(),
+            serde_json::Value::Bool(_) => make::ty("bool"),
+            serde_json::Value::Number(x) => make::ty(if x.is_i64() { "i64" } else { "f64" }),
+            serde_json::Value::String(_) => make::ty("String"),
+            serde_json::Value::Array(x) => {
+                let ty = match x.iter().next() {
+                    Some(x) => self.type_of(x),
+                    None => make::ty_placeholder(),
+                };
+                make::ty(&format!("Vec<{ty}>"))
+            }
+            serde_json::Value::Object(x) => self.build_struct(x),
+        }
+    }
+}
+
+pub(crate) fn json_in_items(acc: &mut Vec<Diagnostic>, file_id: FileId, node: &SyntaxNode) {
+    if node.kind() == SyntaxKind::ERROR
+        && node.first_token().map(|x| x.kind()) == Some(SyntaxKind::L_CURLY)
+        && node.last_token().map(|x| x.kind()) == Some(SyntaxKind::R_CURLY)
+    {
+        let node_string = node.to_string();
```

Comment:
Is there any way to not create a new string here?

Reply:
You could fetch the file's backing string and indexing into that with the node's range